common/argparse.cpp

```diff
@@ -0,0 +1,1362 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/argparse.h"
+
+#include <initializer_list>
+#include <memory>
+
+namespace Carbon {
+
+Args::Arg::Arg(const ArgInfo& info) : info(info) {}
+
+Args::Arg::~Arg() {
+  switch (kind) {
+    case Kind::Flag:
+    case Kind::Integer:
+    case Kind::String:
+    case Kind::MetaActionOnly:
+    case Kind::Invalid:
+      // Nothing to do!
+      break;
+    case Kind::OneOf:
+      value_strings.~decltype(value_strings)();
+      value_action.~ValueActionT();
+      if (has_default) {
+        default_action.~DefaultActionT();
+      }
+      break;
+  }
+}
+
+Args::Command::Command(const CommandInfo& info, Command* parent)
+    : info(info), parent(parent) {}
+
+class Args::MetaPrinter {
+ public:
+  explicit MetaPrinter(llvm::raw_ostream& out) : out_(out) {}
+
+  // Registers this meta printer with a command through the provided builder.
+  //
+  // This adds meta subcommands or options to print both help and version
+  // information for the command.
+  void RegisterWithCommand(const Command& command, CommandBuilder& builder);
+
+  void PrintHelp(const Command& command) const;
+  void PrintVersion(const Command& command) const;
+  void PrintSubcommands(const Command& command) const;
+
+ private:
+  // The indent is calibrated to allow a short and long option after a two
+  // character indent on the prior line to be visually recognized as separate
+  // from the hanging indent.
+  //
+  // Visual guide:                               |  -x, --extract
+  //                                             |          Hanging indented.
+  static constexpr llvm::StringRef BlockIndent = "          ";
+
+  // Width limit for parent command options in usage rendering.
+  static constexpr int MaxParentOptionUsageWidth = 8;
+
+  // Width limit for the leaf command options in usage rendering.
+  static constexpr int MaxLeafOptionUsageWidth = 16;
+
+  void PrintTextBlock(llvm::StringRef indent, llvm::StringRef text) const;
+
+  void PrintRawVersion(const Command& command, llvm::StringRef indent) const;
+  void PrintRawBuildInfo(const Command& command, llvm::StringRef indent) const;
+
+  void PrintArgValueUsage(const Arg& arg) const;
+  void PrintOptionUsage(const Arg& option) const;
+  void PrintShortName(const Arg& arg) const;
+  void PrintArgShortValues(const Arg& arg) const;
+  void PrintArgLongValues(const Arg& arg, llvm::StringRef indent) const;
+  void PrintArgHelp(const Arg& arg, llvm::StringRef indent) const;
+
+  void PrintRawUsageCommandAndOptions(
+      const Command& command,
+      int max_option_width = MaxLeafOptionUsageWidth) const;
+  void PrintRawUsage(const Command& command, llvm::StringRef indent) const;
+  void PrintUsage(const Command& command) const;
+
+  void PrintHelpSubcommands(const Command& command) const;
```

Comment:
It may be worth commenting more of your functions. For example, is this printing subcommands of help, which is my first thought from the name, or help for subcommands, which should maybe be something like "PrintAllSubcommandHelp" instead, or maybe something else?

Reply:
Happy to add comments if useful, but if better names work, that seems even better.

I added some high level grouping comments. Is that enough? Or would a name change been helpful as well? If name change, I'd lean more toward something about the *section* and less about "all". WDYT?